walk/dirinfo.go

```diff
@@ -0,0 +1,151 @@
+package walk
+
+import (
+	"errors"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/bazelbuild/bazel-gazelle/rule"
+)
+
+// dirInfo holds all the information about a directory that Walk2 needs.
+type dirInfo struct {
+	// subdirs and regularFiles hold the names of subdirectories and regular files
+	// that are not ignored or excluded.
+	subdirs, regularFiles []string
+
+	// file is the directory's build file. May be nil if the build file doesn't
+	// exist or contains errors.
+	file *rule.File
+
+	// config is the configuration used by Configurer. We may precompute this
+	// before Configure is called to parallelize directory traversal without
+	// visiting excluded subdirectories.
+	config *walkConfig
+}
+
+// loadDirInfo reads directory info for the directory named by the given
+// slash-separated path relative to the repo root.
+//
+// Do not call this method directly. This should be used with w.cache.get to
+// avoid redundant I/O.
+//
+// loadDirInfo must be called on the parent directory first and the result
+// must be stored in the cache unless rel is "" (repo root).
+//
+// This method may return partial results with an error. For example, if the
+// directory's build file contains a syntax error, the contents of the
+// directory are still returned.
+func (w *walker) loadDirInfo(rel string) (dirInfo, error) {
+	var info dirInfo
+	var errs []error
+	var err error
+	dir := filepath.Join(w.rootConfig.RepoRoot, rel)
+	entries, err := os.ReadDir(dir)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	var parentConfig *walkConfig
+	if rel == "" {
+		parentConfig = getWalkConfig(w.rootConfig)
+	} else {
+		parentRel := path.Dir(rel)
+		if parentRel == "." {
+			parentRel = ""
+		}
+		parentInfo, _ := w.cache.getLoaded(parentRel)
+		parentConfig = parentInfo.config
+	}
+
+	info.file, err = loadBuildFile(parentConfig, w.rootConfig.ReadBuildFilesDir, rel, dir, entries)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	info.config = configureForWalk(parentConfig, rel, info.file)
+
```

Comment:
I think we need `info.config.isExcludedDir(rel)` here to check if the BUILD we just parsed excludes it's own directory 🤯

See https://github.com/bazel-contrib/bazel-gazelle/pull/1928/files#diff-6b93e57c1bff94796acb4cf6b14ec71e6f903b2895bef3e1ab0eb53bb7e9122eR480-R487

Reply:
Done